compiler/rustc_mir_transform/src/check_unsafety.rs

```diff
@@ -251,61 +255,70 @@ impl<'tcx> UnsafetyChecker<'_, 'tcx> {
             .assert_crate_local()
             .lint_root;
         self.register_violations(
-            &[UnsafetyViolation { source_info, lint_root, kind, details }],
-            &[],
+            iter::once(&UnsafetyViolation { source_info, lint_root, kind, details }),
+            iter::empty(),
         );
     }
 
-    fn register_violations(
+    fn register_violations<'a>(
         &mut self,
-        violations: &[UnsafetyViolation],
-        unsafe_blocks: &[(hir::HirId, bool)],
+        violations: impl ExactSizeIterator<Item = &'a UnsafetyViolation>,
```

Comment:
Style nit: AFAICT the properties of `ExactSizeIterator` are not being utilized anywhere in this function, and asking for an iterator here seems to impose both an inconvenient API (e.g. `iter::once` just above) and will also result in multiple monomorphizations of this method.

Am I missing the reason why this couldn't have stayed as a `&[...]`?

Reply:
The `ExactSizeIterator` is used … actually … oh, good point! In 83025b76584f5ed597901df5c7cb1aa974c5da63, there still was one `.len()` call, heh!

Regarding using `&[]` instead, well… the point is the second `new_used_unsafe_blocks` argument that gets passed an iterator over a `HashMap`, so that can’t be a slice anymore. As long as that one’s monomorphized, the first argument doesn’t really matter anymore. On second thought, it might be technically possible to create a new, empty, `HashMap` to iterate over in the other call in `fn require_unsafe`, then avoiding multiple (as in: exactly 2) monomorphizations might become avoidable again.

Reply:
I’ll keep the monomorphization for now. In fact, I think I’d like switching to an `IntoIterator` bound; makes me more convinced that the monomorphized `.for_each` calls can be optimized properly for `iter::once` and `iter::empty`.